tests/test_sbr_calculator.py

```diff
@@ -0,0 +1,524 @@
+        ("test_no_smart_meter", (0.0)),
+    ],
+)
+def test_sbr_score_calculator(user_inputs, expected_sbr_score, request):
+    user_inputs = request.getfixturevalue(user_inputs)
+    sbr_val, _, _ = get_sbr_scores(user_inputs)
+    assert sbr_val == expected_sbr_score
+
+
+@pytest.mark.parametrize(
+    "user_inputs, expected_sbr",
+    [
+        ("test_smartest_home", ("A")),
+        ("test_sbr_b", ("B")),
+        ("test_solar_ev_battery_hp", ("C")),
+        ("test_solar_battery_power_diverter", ("C")),
+        ("test_hp_secondary_heating_hot_water", ("D")),
+        ("test_smart_meter_only", ("G")),
+        ("test_no_smart_meter", ("G")),
+    ],
+)
+def test_sbr_calculator(user_inputs, expected_sbr, request):
+    user_inputs = request.getfixturevalue(user_inputs)
+    _, sbr, _ = get_sbr_scores(user_inputs)
+    assert sbr == expected_sbr
+
+
+@pytest.mark.parametrize(
+    "user_inputs, expected_archetype",
+    [
+        ("test_smartest_home", (FlexArchetype.GOLD_FLEXER)),
+        ("test_sbr_b", (FlexArchetype.GOLD_FLEXER)),
+        ("test_solar_ev_battery_hp", (FlexArchetype.GOLD_FLEXER)),
+        ("test_solar_battery_power_diverter", (FlexArchetype.STRONG_FLEXER)),
+        ("test_hp_secondary_heating_hot_water", (FlexArchetype.GOOD_FLEXER)),
+        ("test_smart_meter_only", (FlexArchetype.LOW_TECH_FLEXER)),
+        ("test_no_smart_meter", (FlexArchetype.UNTAPPED_FLEXER)),
+    ],
+)
+def test_archetype_calculator(user_inputs, expected_archetype, request):
+    user_inputs = request.getfixturevalue(user_inputs)
+    _, _, archetype = get_sbr_scores(user_inputs)
+    assert archetype == expected_archetype
+
+
+def test_all_combinations():
+    """Test valid outputs for all combinations of get_sbr_scores inputs"""
+
+    input_combinations = itertools.product(
+        [True, False],  # smart_meter
+        [True, False],  # smart_ev_charger
+        EVChargerPower.__members__.values(),
+        [True, False],  # smart_v2g_enabled
+        [True, False],  # home_battery
+        BatterySize.__members__.values(),
+        [True, False],  # solar_pv
+        SolarInverterSize.__members__.values(),
+        [True, False],  # electric_heating
+        HeatingSource.__members__.values(),
+        HotWaterSource.__members__.values(),
+        [True, False],  # secondary_heating
+        [True, False],  # secondary_hot_water
+        [True, False],  # integrated_control_sys
+    )
+
+    for inputs in input_combinations:
+        user_inputs = UserInputs(*inputs)
+
+        if (
+            (not inputs[1] and inputs[2] != EVChargerPower.NONE)
+            or (not inputs[4] and inputs[5] != BatterySize.NONE)
+            or (not inputs[6] and inputs[7] != SolarInverterSize.NONE)
+            or (inputs[1] and inputs[2] == EVChargerPower.NONE)
+            or (inputs[4] and inputs[5] == BatterySize.NONE)
+            or (inputs[6] and inputs[7] == SolarInverterSize.NONE)
+        ):
+            with pytest.raises(AssertionError):
```

Comment:
It's unclear exactly what this assertion test is testing for by looking at this code. Looking at the function `sbr_score`, that function is not running any assertion tests.

So it's not clear why I'm expecting certain inputs to be failing here. If assertion tests are failing even further upstream of `sbr_score`, then why are we not testing at source but nesting assertion tests so deeply?

I had to trace all the functions called by `sbr_score` to actually find out assertion tests are carried out in `prep_user_input`. These assertion tests should be written as a test to test `prep_user_input` directly instead of having these tests here.

The whole point of unit tests is so that when test fails, we can quickly identify where the source of bug is. Right now it's not apparent.

Reply:
Makes sense, thanks Sheng. I now test on `prep_user_input` directly